include/cuco/detail/static_map.inl

```diff
@@ -422,6 +422,83 @@ __device__ bool static_map<Key, Value, Scope, Allocator>::device_mutable_view::i
   }
 }
 
+template <typename Key, typename Value, cuda::thread_scope Scope, typename Allocator>
+template <typename Hash, typename KeyEqual>
+__device__
+  thrust::pair<typename static_map<Key, Value, Scope, Allocator>::device_mutable_view::iterator,
+               bool>
+  static_map<Key, Value, Scope, Allocator>::device_mutable_view::fetch_insert(
+    value_type const& insert_pair, Hash hash, KeyEqual key_equal) noexcept
+{
+  auto current_slot{initial_slot(insert_pair.first, hash)};
+
+  while (true) {
+    key_type const existing_key = current_slot->first.load(cuda::std::memory_order_relaxed);
+    // The user provide `key_equal` can never be used to compare against `empty_key_sentinel` as the
+    // sentinel is not a valid key value. Therefore, first check for the sentinel
+    auto const slot_is_available =
+      detail::bitwise_compare(existing_key, this->get_empty_key_sentinel()) or
+      detail::bitwise_compare(existing_key, this->get_erased_key_sentinel());
+
+    // the key we are trying to insert is already in the map, so we return with failure to insert
+    if (not slot_is_available and key_equal(existing_key, insert_pair.first)) {
+      // If we cannot use a single CAS operation, ensure that the write to
+      // the value part also took place.
+      if constexpr (not cuco::detail::is_packable<value_type>()) {
+        auto& slot_value       = current_slot->second;
+        auto const empty_value = this->get_empty_value_sentinel();
+        while (
+          detail::bitwise_compare(slot_value.load(cuda::std::memory_order_relaxed), empty_value)) {
+          // spin
+        }
+      }
```

Comment:
I'm not entirely convinced that this is safe if we take deletions into account.
The current spin loop relies on the value slot only being able to transition from empty->filled but a deletion triggers filled->empty. So maybe we need to disable this functionality if deletions are enabled.

Reply:
I *think* this is fine because it shouldn't be possible to write legitimate code that concurrently deletes a key and modifies its value. Even if this was made to work here, user code still wouldn't be allowed to dereference an iterator because another thread might have erased it. Probably a good idea to document this; also users must not mix `insert` and `fetch_insert` because the former could still fall back to `back_to_back_cas`.

Reply:
Ah, yes, you're right! The constrains that apply to concurrent `ìnsert`+`erase` also apply to `fetch_insert`+`erase`.

> also users must not mix insert and fetch_insert because the former could still fall back to back_to_back_cas.

If we disable `fetch_insert` for pre-Volta architectures, then there should be no issue in using `insert`+`fetch_insert` concurrently. A simple `#if (CUCO_CUDA_MINIMUM_ARCH >=700)` should do the trick. BTW this macro is introduced by #201 so you might need to cherry-pick those commits first.

Reply:
> > also users must not mix insert and fetch_insert because the former could still fall back to back_to_back_cas.
> 
> If we disable `fetch_insert` for pre-Volta architectures, then there should be no issue in using `insert`+`fetch_insert` concurrently. A simple `#if (CUCO_CUDA_MINIMUM_ARCH >=700)` should do the trick. BTW this macro is introduced by #201 so you might need to cherry-pick those commits first.

Unfortunately, it's exactly the other way around: pre-Volta uses `cas_dependent_write` and we likely can't support `fetch_insert` anyway; Volta+ uses `back_to_back_cas` for `insert` (for performance reasons, but I personally don't see much of a difference on my Turing card), which you must not mix with `fetch_insert`.

Reply:
Dang, sorry for my confusion :D Well, in that case, I agree that we should add a section about this in the documentation.